### ISSUE TITLE
Option to vertically format arguments if wrapped.

### ISSUE THREAD

**Issue by funkybob** [astral-sh/ruff#9992]
Would it be possible to add a formatter option so that once it decides to split arguments to a call onto a new line, each argument goes on its own line?

So, for example, when it chooses to reformat:
```
foo = bar(long_name=value, another_argument=1, something_else="12314")
```
into:
```
foo = bar(
    long_name=value, another_argument=1, something_else="12314"
)
```
it will instead produce:
```
foo = bar(
    long_name=value,
    another_argument=1,
    something_else="12314"
)
```
(and ideally also add the trailing comma, but that's a separate formatting option, I assume)

I know I can force this formatting to be retained by using the magic trailing comma, but I'd like to not have to think about this.

**Comment by MichaReiser** [comment]
Hy @funkybob 

The main request in this issue is to have an option to disable the style where ruff tries to format all arguments on a single line before splitting each argument on a separate line. Ruff should already add the trailing comma as you want it to: It adds the trailing comma when each argument is on its own line, but it omits it when formatting all arguments on a single line (the style that you don't want to have).

We're still considering our position on how configurable our formatter should be. There are arguments for an opinionated formatter with very few options. But supporting many options also has its benefits. For now, our position is to avoid formatting style-related options, which includes the option you're asking for.

**Comment by benblank** [comment]
If ruff prefers to avoid configurable formatting, then I think there's a case to be made that "one argument per line" should be how calls like this are wrapped.

Basically, the point of wrapping code is to improve readability, right? Long, unbroken lines may extend past the edge of a reading area and also tend to be more difficult to parse visually. Good wrapping helps with both of those issues.

But this kind of "function name, then every-argument-on-a-single-line, then close parenthesis" wrapping doesn't really do a lot for line lengths. This is especially noticeable with short function names; in funkybob's second example, the indented arguments line is only seven characters shorter than the original! A few more keystrokes, and it'll need rewrapped anyway.

I also don't think it does hardly anything for visual parsing; again, looking at the examples above, it's *really easy* to quickly spot that there are three arguments in the last example. But I find myself having to sort of stop-and-count-the-equals-signs in the other two. Putting each argument on a separate line makes a big difference for readability, I believe.

Finally, there's also a sense in which it seems oddly inconsistent to me. It feels like there's a kind of "gap" between calls which are short enough to not need wrapped at all and those which are long enough that all the arguments can't be made to fit on a single line, anyway. In between the two, there's this kind of "partial wrapping" which triples the vertical space taken by the call but often doesn't significantly reduce the horizontal space, nor improve readability much.

**TL;DR:** If ruff is going to have only one wrapping behavior in this scenario, I think it should be the one funkybob describes, not it's current behavior… even though it would be a breaking change. 😓

**Comment by alexei** [comment]
Because sometimes ruff puts each argument on its own line, and sometimes it puts all arguments on the same line, I often find myself having to review the changes and make manual adjustments. I too support to either have an option to instruct ruff to lean on always adding the comma, or have that as the default behavior. This would help enforcing a consistent style, especially in larger brownfield projects.

**Comment by jorenham** [comment]
I wrote a simple `libcst` codemod that adds a trailing comma to such half-wrapped parameters, so that `ruff format` will place them onto separate lines after that:

https://github.com/jorenham/scipy-stubs/blob/dd90c64c05a8f75f684662c277278e2c35400d4f/codegen/mods.py#L97-L121

But clearly this is far from ideal; as it requires running `ruff format` twice, and doesn't work if `skip-magic-trailing-comma` is set.

**Comment by Tatsh** [comment]
I would like to have the option to keep arguments aligned after the bracket:

```python
func(long1,
     long2,
     long3)
```

YAPF has an implicit way of controlling the line breaks which is by adding a trailing `,\n)` to the list or function arguments:

```python
foo(1, 2, 3,
)
```

becomes:

```python
foo(
    1,
    2,
    3,
)
``

These are what are keeping me from switching. Please make it configurable especially since YAPF is seeming more and more like a dead project.

**Comment by Avasam** [comment]
> disable the style where ruff tries to format all arguments on a single line before splitting each argument on a separate line

Basically. Others have already said it in more details, but for me:

1. I see this current behaviour as going against the idea of formatting code so that it reduces merge conflicts. (you now have wrapping style transitions instead of one).
2. It's also not any more readable. Either fit on a line, or list the parameters, call args, comprehension statements (ref https://github.com/astral-sh/ruff/issues/11753), etc. individually.

**Comment by Avasam** [comment]
> I wrote a simple `libcst` codemod that adds a trailing comma to such half-wrapped parameters, so that `ruff format` will place them onto separate lines after that:
> 
> [jorenham/scipy-stubs@`dd90c64`/codegen/mods.py#L97-L121](https://github.com/jorenham/scipy-stubs/blob/dd90c64c05a8f75f684662c277278e2c35400d4f/codegen/mods.py#L97-L121)

Your codemod is pretty small. But I'll still point to existing work https://pypi.org/project/add-trailing-comma/ which I have used. (I appreciate your work on scipy-stubs btw!)

**Comment by jorenham** [comment]
> > I wrote a simple `libcst` codemod that adds a trailing comma to such half-wrapped parameters, so that `ruff format` will place them onto separate lines after that:
> > [jorenham/scipy-stubs@`dd90c64`/codegen/mods.py#L97-L121](https://github.com/jorenham/scipy-stubs/blob/dd90c64c05a8f75f684662c277278e2c35400d4f/codegen/mods.py#L97-L121)
> 
> Your codemod is pretty small. But I'll still point to existing work [pypi.org/project/add-trailing-comma](https://pypi.org/project/add-trailing-comma/) which I have used. 
Oh I didn't know about this one &mdash; I'll give it try. Thanks for sharing

> (I appreciate your work on scipy-stubs btw!)
That's good to to hear! Thank you for contributing to it, as well :)

**Comment by PedroPerpetua** [comment]
Seconding this issue; I'd love to have the ability to enforce this!

**Comment by ivangabriele** [comment]
@funkybob Maybe we should retitle / update this proposal to "Always align arguments vertically for any wrapped statement"?

**Comment by jorenham** [comment]
I would prefer this be an extra option, similar to the [`trailing_comma`](https://rust-lang.github.io/rustfmt/?version=v1.8.0&search=#trailing_comma) option of `rustfmt`, which can be either `Vertical` (the current behavior), `Always` (the new option), or `Never` (if disabled)

**Comment by ivangabriele** [comment]
> I would prefer this be an extra option, similar to the [`trailing_comma`](https://rust-lang.github.io/rustfmt/?version=v1.8.0&search=#trailing_comma) option of `rustfmt`, which can be either `Vertical` (the current behavior), `Always` (the new option), or `Never` (if disabled)

I may be mistaken but I think the point of this proposal is to avoid "intermediate" states with some wrapping cases keeping args horizontally aligned. In Rust, by default, as soon as a list of args forces a line to wrap, they are aligned vertically without any "intermediary" state.

Rust code example with a classic 80 line width:

```rust
pub async fn get_dashboard_state(app_handle: AppHandle, shared_state: State<'_, state::DashboardSharedState>) -> Result<(), ()> {
```

is auto-formatted like that by `rustfmt`:

```rust
pub async fn get_dashboard_state(
    app_handle: AppHandle,
    shared_state: State<'_, state::DashboardSharedState>,
) -> Result<(), ()> {
```

and **not** like that:

```rust
pub async fn get_dashboard_state(
    app_handle: AppHandle, shared_state: State<'_, state::DashboardSharedState>,
) -> Result<(), ()> {
```

which IMO makes it a lot more readable.

**Comment by funkybob** [comment]
@ivangabriele you are correct.

I find the intermediate step pointless, as it really doesn't improve readability, it just prevents a line being over the max line length.

In all honesty I would prefer the choice to split to arg-per-line not depend on line length; however, I'll be the first to concede that trying to determine a rule for when readability is impacted is far from trivial.

**Comment by nicknicolini** [comment]
@MichaReiser @ntBre i'm curious if there is an up to date viewpoint on this ticket, I think @ivangabriele has the best articulation of what is being asked for [here](https://github.com/astral-sh/ruff/issues/9992#issuecomment-2975063491). Was wondering if ruff devs are open to putting the behavior behind a flag and/or changing the behavior in the next formatting break

As it stands, ruff users are forced to choose being using ruff with imo less readable "intermediary" wrapped code (as outlined by Ivan above), or using a separate, slower, formatting tool in pre-commit just to achieve desired vertical formatting (e.g. `add-trailing-commas`). Both of these options are kind of a bummer coming from formatting tooling in other languages

**Comment by MichaReiser** [comment]
No, there's no update on this issue. This formatting is also not specific to arguments. The same 3-layout formatting exists for many other nodes, e.g. dictionaries and an option would have to address all of them holistically or the formatting style becomes inconsistent.

**Comment by funkybob** [comment]
If it makes any difference, if I saw the same thing happen on dicts (moving the items to a new line to avoid line limit) I'd also object.

If it's hard to read on one line, it's hard to read on one line; it should be split to many lines.

**Comment by nicknicolini** [comment]
> No, there's no update on this issue. This formatting is also not specific to arguments. The same 3-layout formatting exists for many other nodes, e.g. dictionaries and an option would have to address all of them holistically or the formatting style becomes inconsistent.

@MichaReiser does that mean ruff doesn't want to make a decision here? or could a decision be made + OSS contribution on the table? i'd imagine folks on this ticket would be willing to contribute (myself included) if there's a clear decision

**Comment by MichaReiser** [comment]
> does that mean ruff doesn't want to make a decision here?

No, it's just not a priority for us right now. We're focusing on features with higher demand, but we might pick this up when working on the 2026 formatter style (which is when someone has time to focus on the formatter).

>  or could a decision be made + OSS contribution on the table?

That would certainly help drive the discussion forward but it probably still has the outcome that we won't have time to make the decision before working on the new formatter style.

**Comment by nicknicolini** [comment]
> > does that mean ruff doesn't want to make a decision here?
> 
> No, it's just not a priority for us right now. We're focusing on features with higher demand, but we might pick this up when working on the 2026 formatter style (which is when someone has time to focus on the formatter).
> 
> > or could a decision be made + OSS contribution on the table?
> 
> That would certainly help drive the discussion forward but it probably still has the outcome that we won't have time to make the decision before working on the new formatter style.

sounds good, just wanted to confirm! If a decision is made on what ruff would like to support and you are open to a contribution for the change, feel free to give an update here whenever that happens. Would be willing to take a stab at it

**Comment by ivangabriele** [comment]
> No, it's just not a priority for us right now. We're focusing on features with higher demand, but we might pick this up when working on the 2026 formatter style (which is when someone has time to focus on the formatter).

@MichaReiser I ask just in case: would a PR help or would it be useless right now regarding your roadmap? It may be non-trivial (I didn't take time to check how that part works yet) and considered a breaking change (even though we can always make that an opt-in option on the current version).

**Comment by MichaReiser** [comment]
@ivangabriele A PR could help to assess the changes but, as stated above, I don't think will get to reviewing or making a decision on this before working on the 2026 style guide.
